File: go-legacy/internal/cli/migrate.go
// https://hypercall.xyz

package cli

import (
	"encoding/json"
	"flag"
	"fmt"
	"os"
	"path/filepath"

	"github.com/jake/mcpzip/internal/config"
	"github.com/jake/mcpzip/internal/types"
)

func runMigrate(args []string) error {
	fs := flag.NewFlagSet("migrate", flag.ContinueOnError)
	outputPath := fs.String("config", config.DefaultPath(), "output config file path")
	claudeConfigPath := fs.String("claude-config", "", "path to Claude Code config (auto-detected if empty)")
	dryRun := fs.Bool("dry-run", false, "show what would happen without writing files")
	if err := fs.Parse(args); err != nil {
		return err
	}

	// Find the Claude Code config path.
	claudePath := *claudeConfigPath
	if claudePath == "" {
		var err error
		claudePath, err = config.FindClaudeCodeConfigPath()
		if err != nil {
			return fmt.Errorf("finding Claude Code config: %w", err)
		}
	}

	// Load Claude Code config.
	claudeCfg, err := config.LoadClaudeCodeConfigFrom(claudePath)
	if err != nil {
		return fmt.Errorf("loading Claude Code config: %w", err)
	}

	// Find the mcpzip binary path for the Claude Code config entry.
	mcpzipBin, err := os.Executable()
	if err != nil {
		mcpzipBin = "mcpzip"
	}

	if *dryRun {
		fmt.Printf("Dry run: would migrate %d server(s) from %s\n\n", len(claudeCfg.MCPServers), claudePath)
		fmt.Printf("1. Write mcpzip config to %s:\n", *outputPath)
		for name, sc := range claudeCfg.MCPServers {
			fmt.Printf("   - %s (%s)\n", name, sc.EffectiveType())
		}
		fmt.Printf("\n2. Update %s:\n", claudePath)
		fmt.Printf("   - Remove %d individual server entries\n", len(claudeCfg.MCPServers))
		fmt.Printf("   - Add single \"mcpzip\" entry pointing to %s\n", mcpzipBin)
		return nil
	}

	// Step 1: Write the mcpzip proxy config.
	if err := writeProxyConfig(claudeCfg, *outputPath); err != nil {
		return err
	}

	// Step 2: Update the Claude Code config.
	if err := updateClaudeConfig(claudePath, claudeCfg, mcpzipBin); err != nil {
		return err
	}

	fmt.Printf("\nDone! Restart Claude Code to use mcpzip.\n")
	return nil
}

// writeProxyConfig writes the mcpzip config file with all the migrated servers.
func writeProxyConfig(claudeCfg *config.ClaudeCodeConfig, outputPath string) error {
	proxyCfg := types.ProxyConfig{
		MCPServers: make(map[string]types.ServerConfig, len(claudeCfg.MCPServers)),
	}
	for name, sc := range claudeCfg.MCPServers {
		proxyCfg.MCPServers[name] = sc
	}

	data, err := json.MarshalIndent(proxyCfg, "", "  ")
	if err != nil {
		return fmt.Errorf("marshaling config: %w", err)
	}
	data = append(data, '\n')

	dir := filepath.Dir(outputPath)
	if err := os.MkdirAll(dir, 0755); err != nil {
		return fmt.Errorf("creating config directory: %w", err)
	}

	if err := os.WriteFile(outputPath, data, 0600); err != nil {
		return fmt.Errorf("writing config: %w", err)
	}

	fmt.Printf("Wrote mcpzip config to %s (%d servers)\n", outputPath, len(proxyCfg.MCPServers))
	for name, sc := range proxyCfg.MCPServers {
		fmt.Printf("  - %s (%s)\n", name, sc.EffectiveType())
	}
	return nil
}

// updateClaudeConfig replaces all mcpServers in the Claude Code config with
// a single mcpzip entry. Preserves all other fields in the config.
func updateClaudeConfig(claudePath string, claudeCfg *config.ClaudeCodeConfig, mcpzipBin string) error {
	// Read the raw JSON to preserve all non-mcpServers fields.
	data, err := os.ReadFile(claudePath)
	if err != nil {
		return fmt.Errorf("reading %s: %w", claudePath, err)
	}

	var raw map[string]json.RawMessage
	if err := json.Unmarshal(data, &raw); err != nil {
		return fmt.Errorf("parsing %s: %w", claudePath, err)
	}

	// Replace mcpServers with single mcpzip entry.
	newServers := map[string]types.ServerConfig{
		"mcpzip": {
			Type:    "stdio",
			Command: mcpzipBin,
			Args:    []string{"serve"},
		},
	}
	serversJSON, err := json.Marshal(newServers)
	if err != nil {
		return fmt.Errorf("marshaling new mcpServers: %w", err)
	}
	raw["mcpServers"] = serversJSON

	out, err := json.MarshalIndent(raw, "", "  ")
	if err != nil {
		return fmt.Errorf("marshaling updated config: %w", err)
	}
	out = append(out, '\n')

	if err := os.WriteFile(claudePath, out, 0600); err != nil {
		return fmt.Errorf("writing %s: %w", claudePath, err)
	}

	fmt.Printf("\nUpdated %s:\n", claudePath)
	fmt.Printf("  Replaced %d servers with single mcpzip entry\n", len(claudeCfg.MCPServers))
	return nil
}

File: src/cli/migrate.rs
use std::collections::HashMap;
use std::path::{Path, PathBuf};

use clap::Args;

use crate::config;
use crate::error::McpzipError;
use crate::types::{ProxyConfig, ServerConfig};

#[derive(Args)]
pub struct MigrateArgs {
    /// Output config file path
    #[arg(long, default_value_os_t = config::default_path())]
    pub config: PathBuf,

    /// Path to Claude Code config (auto-detected if empty)
    #[arg(long)]
    pub claude_config: Option<PathBuf>,

    /// Show what would happen without writing files
    #[arg(long)]
    pub dry_run: bool,
}

pub fn run_migrate(args: &MigrateArgs) -> Result<(), McpzipError> {
    // Find Claude Code config.
    let claude_path = match &args.claude_config {
        Some(p) => p.clone(),
        None => config::find_claude_code_config_path()?,
    };

    // Load Claude Code config.
    let claude_cfg = config::load_claude_code_config_from(&claude_path)?;

    // Find the mcpzip binary path.
    let mcpzip_bin = std::env::current_exe()
        .map(|p| p.to_string_lossy().into_owned())
        .unwrap_or_else(|_| "mcpzip".into());

    if args.dry_run {
        println!(
            "Dry run: would migrate {} server(s) from {}",
            claude_cfg.mcp_servers.len(),
            claude_path.display()
        );
        println!("\n1. Write mcpzip config to {}:", args.config.display());
        for (name, sc) in &claude_cfg.mcp_servers {
            println!("   - {} ({})", name, sc.effective_type());
        }
        println!("\n2. Update {}:", claude_path.display());
        println!(
            "   - Remove {} individual server entries",
            claude_cfg.mcp_servers.len()
        );
        println!(
            "   - Add single \"mcpzip\" entry pointing to {}",
            mcpzip_bin
        );
        return Ok(());
    }

    // Step 1: Write the mcpzip proxy config.
    write_proxy_config(&claude_cfg.mcp_servers, &args.config)?;

    // Step 2: Update the Claude Code config.
    update_claude_config(&claude_path, claude_cfg.mcp_servers.len(), &mcpzip_bin)?;

    println!("\nDone! Restart Claude Code to use mcpzip.");
    Ok(())
}

/// Write the mcpzip config file with all migrated servers.
pub fn write_proxy_config(
    servers: &HashMap<String, ServerConfig>,
    output_path: &Path,
) -> Result<(), McpzipError> {
    let proxy_cfg = ProxyConfig {
        gemini_api_key: None,
        search: Default::default(),
        idle_timeout_minutes: None,
        call_timeout_seconds: None,
        mcp_servers: servers.clone(),
    };

    let data = serde_json::to_string_pretty(&proxy_cfg)?;

    if let Some(parent) = output_path.parent() {
        std::fs::create_dir_all(parent)?;
    }
    std::fs::write(output_path, format!("{}\n", data))?;

    println!(
        "Wrote mcpzip config to {} ({} servers)",
        output_path.display(),
        servers.len()
    );
    for (name, sc) in servers {
        println!("  - {} ({})", name, sc.effective_type());
    }
    Ok(())
}

/// Replace mcpServers in Claude Code config with a single mcpzip entry.
fn update_claude_config(
    claude_path: &Path,
    old_count: usize,
    mcpzip_bin: &str,
) -> Result<(), McpzipError> {
    let data = std::fs::read_to_string(claude_path)?;
    let mut raw: serde_json::Map<String, serde_json::Value> = serde_json::from_str(&data)?;

    // Replace mcpServers with single mcpzip entry.
    let mut new_servers = serde_json::Map::new();
    new_servers.insert(
        "mcpzip".into(),
        serde_json::json!({
            "type": "stdio",
            "command": mcpzip_bin,
            "args": ["serve"]
        }),
    );
    raw.insert("mcpServers".into(), serde_json::Value::Object(new_servers));

    let out = serde_json::to_string_pretty(&raw)?;
    std::fs::write(claude_path, format!("{}\n", out))?;

    println!("\nUpdated {}:", claude_path.display());
    println!("  Replaced {} servers with single mcpzip entry", old_count);
    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_write_proxy_config_basic() {
        let dir = tempfile::tempdir().unwrap();
        let output = dir.path().join("config.json");

        let mut servers = HashMap::new();
        servers.insert(
            "slack".into(),
            ServerConfig {
                server_type: None,
                command: Some("slack-mcp".into()),
                args: Some(vec!["--token".into(), "abc".into()]),
                env: None,
                url: None,
                headers: None,
            },
        );
        servers.insert(
            "github".into(),
            ServerConfig {
                server_type: None,
                command: Some("gh-mcp".into()),
                args: None,
                env: None,
                url: None,
                headers: None,
            },
        );

        write_proxy_config(&servers, &output).unwrap();

        let data = std::fs::read_to_string(&output).unwrap();
        let proxy_cfg: ProxyConfig = serde_json::from_str(&data).unwrap();
        assert_eq!(proxy_cfg.mcp_servers.len(), 2);
        assert_eq!(
            proxy_cfg.mcp_servers["slack"].command.as_deref(),
            Some("slack-mcp")
        );
    }

    #[test]
    fn test_write_proxy_config_empty() {
        let dir = tempfile::tempdir().unwrap();
        let output = dir.path().join("config.json");

        let servers = HashMap::new();
        write_proxy_config(&servers, &output).unwrap();

        let data = std::fs::read_to_string(&output).unwrap();
        let proxy_cfg: ProxyConfig = serde_json::from_str(&data).unwrap();
        assert_eq!(proxy_cfg.mcp_servers.len(), 0);
    }

    #[test]
    fn test_write_proxy_config_creates_directory() {
        let dir = tempfile::tempdir().unwrap();
        let output = dir.path().join("nested").join("deep").join("config.json");

        let mut servers = HashMap::new();
        servers.insert(
            "test".into(),
            ServerConfig {
                server_type: None,
                command: Some("test-mcp".into()),
                args: None,
                env: None,
                url: None,
                headers: None,
            },
        );

        write_proxy_config(&servers, &output).unwrap();
        assert!(output.exists());
    }

    #[test]
    fn test_write_proxy_config_preserves_http() {
        let dir = tempfile::tempdir().unwrap();
        let output = dir.path().join("config.json");

        let mut servers = HashMap::new();
        servers.insert(
            "remote".into(),
            ServerConfig {
                server_type: Some("http".into()),
                command: None,
                args: None,
                env: None,
                url: Some("http://localhost:8080/mcp".into()),
                headers: None,
            },
        );

        write_proxy_config(&servers, &output).unwrap();

        let data = std::fs::read_to_string(&output).unwrap();
        let proxy_cfg: ProxyConfig = serde_json::from_str(&data).unwrap();
        let s = &proxy_cfg.mcp_servers["remote"];
        assert_eq!(s.server_type.as_deref(), Some("http"));
        assert_eq!(s.url.as_deref(), Some("http://localhost:8080/mcp"));
    }

    #[test]
    fn test_update_claude_config() {
        let dir = tempfile::tempdir().unwrap();
        let claude_path = dir.path().join("claude.json");

        let content = r#"{"mcpServers": {"test": {"command": "test-mcp"}}}"#;
        std::fs::write(&claude_path, content).unwrap();

        update_claude_config(&claude_path, 1, "/usr/local/bin/mcpzip").unwrap();

        let data = std::fs::read_to_string(&claude_path).unwrap();
        let raw: serde_json::Map<String, serde_json::Value> = serde_json::from_str(&data).unwrap();
        let servers = raw["mcpServers"].as_object().unwrap();
        assert_eq!(servers.len(), 1);
        assert!(servers.contains_key("mcpzip"));
        assert_eq!(servers["mcpzip"]["command"], "/usr/local/bin/mcpzip");
    }
}
